src/main/java/htsjdk/samtools/SAMSequenceRecord.java

```diff
@@ -98,68 +107,185 @@ public SAMSequenceRecord(final String name, final int sequenceLength) {
+    public String getSpecies() {
+        return (String) getAttribute(SPECIES_TAG);
+    }
+
+    public void setSpecies(final String value) {
+        setAttribute(SPECIES_TAG, value);
+    }
+
+    public String getMd5() {
+        return (String) getAttribute(MD5_TAG);
+    }
+
+    public void setMd5(final String value) {
+        setAttribute(MD5_TAG, value);
+    }
+
+    public String getDescription() {
+        return getAttribute(DESCRIPTION_TAG);
+    }
+
+    public void setDescription(final String value) {
+        setAttribute(DESCRIPTION_TAG, value);
+    }
 
     /**
-     * @return Index of this record in the sequence dictionary it lives in. 
+     * @return Index of this record in the sequence dictionary it lives in.
      */
-    public int getSequenceIndex() { return mSequenceIndex; }
+    public int getSequenceIndex() {
+        return mSequenceIndex;
+    }
 
     // Private state used only by SAM implementation.
-    public void setSequenceIndex(final int value) { mSequenceIndex = value; }
+    public void setSequenceIndex(final int value) {
+        mSequenceIndex = value;
+    }
+
+    /**
+     * Returns unmodifiable set with alternative sequence names.
+     */
+    public Set<String> getAlternativeSequenceNames() {
+        final String anTag = getAttribute(ALTERNATIVE_SEQUENCE_NAME_TAG);
+        return (anTag == null) ? Collections.emptySet()
+                : Collections.unmodifiableSet(new HashSet<>(Arrays.asList(anTag.split(ALTERNATIVE_SEQUENCE_NAME_SEPARATOR))));
+    }
+
+    /**
+     * Adds an alternative sequence name if it is not the same as the sequence name or it is not present already.
+     */
+    public void addAlternativeSequenceName(final String name) {
+        validateAltRegExp(name);
+        final Set<String> altSequences = new HashSet<>(getAlternativeSequenceNames());
+
+        if (!mSequenceName.equals(name)) {
+            altSequences.add(name);
+        }
+        encodeAltSequences(altSequences);
+    }
+
+    /**
+     * Sets the alternative sequence names in the order provided by iteration, removing the previous values.
+     */
+    public void setAlternativeSequenceName(final Collection<String> alternativeSequences) {
+        if (alternativeSequences == null) {
+            setAttribute(ALTERNATIVE_SEQUENCE_NAME_TAG, null);
+        } else {
+            // validate all the names and encode afterwards
+            alternativeSequences.forEach(SAMSequenceRecord::validateAltRegExp);
+            encodeAltSequences(alternativeSequences);
+        }
+    }
+
+    private static void validateAltRegExp(final String name) {
+        if (!LEGAL_RNAME_PATTERN.matcher(name).matches()) {
+            throw new IllegalArgumentException(String.format("Invalid alternative sequence name '%s': do not match the pattern %s", name, LEGAL_RNAME_PATTERN));
+        }
+    }
+
+    private void encodeAltSequences(final Collection<String> alternativeSequences) {
```

Comment:
I think you could probably save a bunch of code/ potential error paths if the error checking / deduplication is all done here.

Something like:

encoded ==   alternativeSequences.isEmpty() ? null : alternativeSequences.stream()
                .sorted()
                .distinct()
                .peek(SAMSequenceRecord::validateAltRegExp)
                .collect(Collectors.joining(ALTERNATIVE_SEQUENCE_NAME_SEPARATOR));

Reply:
I did it...but I'm not sure what code you think I can remove now....

Reply:
I thought you could remove the various guards in the code that call into this and let this be the single source of error checking.